backend: replace BIP69 with random sorting for tx inputs/outputs

strmci opened the pull request:
This commit refactors the sorting mechanism for transaction inputs and outputs. It replaces the existing BIP69 lexicographical sorting with a randomized sorting approach.

BIP69, or "Lexicographic Order of Transaction Inputs and Outputs," was initially proposed to enhance privacy and mitigate blockchain fingerprinting. However, due to its limited adoption and effectiveness, it has been replaced by random shuffling of inputs and outputs, offering improved privacy benefits.

Beerosagos commented on backend/coins/btc/maketx/maketx.go:
```diff
@@ -168,7 +168,11 @@ func NewTxSpendAll(
-	txsort.InPlaceSort(unsignedTransaction)
+
+	// Shuffle inputs and outputs
+	ShuffleTxInputs(unsignedTransaction)
```
small nit: looks like you don't really need two separate functions, right?

strmci replied:
I agree, having just one function is sufficient, I've made the update, PTAL

benma commented on backend/coins/btc/maketx/maketx.go:
```diff
@@ -263,3 +267,16 @@ func NewTx(
 }
+
+// ShuffleTxInputsAndOutputs shuffles both the TxIn and TxOut slices of a wire.MsgTx.
+func ShuffleTxInputsAndOutputs(tx *wire.MsgTx) {
```
can unexport this `shuffleTxInputsAndOutputs` as it is only used locally.

benma reviewed (commented):
Would be great if you could explain the reason for this in the commit msg

Beerosagos reviewed (approved):
tACK!

Would be nice to have a test, but I don't think there is a good way to check that the ins and outs are actually shuffled. wdyt @benma ?

benma commented:
> tACK!
> 
> Would be nice to have a test, but I don't think there is a good way to check that the ins and outs are actually shuffled. wdyt @benma ?

test would be nice, and you can do it by using a local `rand.Rand` object which is initialized with a constant seed in tests.

strmci commented:
@Beerosagos @benma  I made some updates:
- updated the `ShuffleTxInputsAndOutputs` so it uses the `rand.Rand` object and added unit test
- kept it exported as it is used in the test file
- updated changelog file
PTAL

Beerosagos commented on backend/coins/btc/maketx/maketx.go:
```diff
@@ -168,7 +171,10 @@ func NewTxSpendAll(
 	}
-	txsort.InPlaceSort(unsignedTransaction)
+
+	secureRand := mrand.New(mrand.NewSource(secureSeed()))
```
I'm not 100% sure, but I think that it would be better to seed the generator using a random int64 obtained from the default rand generator.  @benma wdyt?

benma commented on backend/coins/btc/maketx/maketx_test.go:
```diff
@@ -312,3 +313,40 @@ func (s *newTxSuite) TestNewTxCoinSelection() {
+	tx.TxOut = txOuts
+	// Shuffle with constant seed
+	testRand := rand.New(rand.NewSource(1000))
+	maketx.ShuffleTxInputsAndOutputs(tx, testRand)
```
change the first line from `package maketx_test` to `package maketx`, then you can access the unexported stuff. this way, you can unexport `ShuffleTxInputsAndOutputs()` again (it should not be exported as it is only used internally).

benma reviewed (changes requested):
utACK, with a small nit about unexporting the function again

benma commented:
> also please add a note to CHANGELOG.md 😄

@strmci also this ^

strmci commented:
@benma  

- as we discussed, I created a separate file `maketx_internal_test.go` to keep the function unexported
- CHANGELOG.md was already updated

PTAL

benma reviewed (approved):
LGTM
